mlir/lib/Catalyst/Transforms/TBAAPatterns.cpp

```diff
@@ -22,16 +23,51 @@
 #include "mlir/Dialect/LLVMIR/LLVMDialect.h"
 
 #include "Catalyst/Transforms/TBAAUtils.h"
+#include "llvm/Support/Casting.h"
 
 using namespace mlir;
 
 namespace catalyst {
-void setTag(mlir::Type baseType, catalyst::TBAATree *tree, mlir::MLIRContext *ctx,
+
+bool isFromExtractAlignedPointerAsIndexOp(Operation *op)
+{
+    if (isa<memref::ExtractAlignedPointerAsIndexOp>(op)) {
+        return true;
+    }
+    auto prevOp = op->getPrevNode();
+    if (prevOp) {
+        return isFromExtractAlignedPointerAsIndexOp(prevOp);
+    }
+    return false;
+}
+
+bool isInsideDeallocHelper(Operation *op)
+{
+    auto parentOp = dyn_cast<func::FuncOp>(op->getParentOp());
+    if (parentOp) {
+        auto str = parentOp.getName();
+        llvm::StringRef dellocRef = "dealloc_helper";
+        return str.compare(dellocRef) == 0;
+    }
+
+    llvm::outs() << parentOp.getName() << "\n";
+    return false;
+}
+
+void setTag(mlir::Type baseType, catalyst::TBAATree *tree, Operation *currentOp,
             mlir::LLVM::AliasAnalysisOpInterface newOp)
 {
+    mlir::MLIRContext *ctx = currentOp->getContext();
     mlir::LLVM::TBAATagAttr tag;
     if (isa<IndexType>(baseType) || isa<IntegerType>(baseType)) {
-        tag = tree->getTag("int");
+        // Index can be used as a pointer.
+        if (isa<IndexType>(baseType) &&
+            (isFromExtractAlignedPointerAsIndexOp(currentOp) || isInsideDeallocHelper(currentOp))) {
```

Comment:
Can you extract the logic `isFromExtractAlignedPointerAsIndexOp(currentOp) || isInsideDeallocHelper(currentOp)` from set tags? You should be able to determine the type before the setting it.

Reply:
Yes, I have moved this logic out.